Adds Sidecar API (#19)
moondance-labs/tanssi-docs: builders/interact/substrate-api/sidecar-api.md

## Installing and Running Substrate API Sidecar {: #installing-and-running-substrate-api-sidecar } 

There are multiple ways of installing and running the Substrate API Sidecar. This guide will describe the steps for installing and running it locally through NPM. For running Substrate API Sidecar through Docker, or building and running it from source, please refer to the [Substrate API Sidecar Github Repository](https://github.com/paritytech/substrate-api-sidecar#readme).

--8<-- 'text/common/node-npm.md'

### Installing the Substrate API Sidecar {: #installing-the-substrate-api-sidecar }

To install the Substrate API Sidecar service locally in the current directory, run this from the command line:

```
npm install @substrate/api-sidecar@{{ networks.dancebox.substrate_api_sidecar.stable_version }}
```

!!! note
    If the current folder does not already have a Node.js project structure, you need to manually created the `node_modules` directory by typing `mkdir node_modules`.

> Substrate API Sidecar v{{ networks.dancebox.substrate_api_sidecar.stable_version }} is the current stable version that has been tested to work with ContainerChains. You can verify the installation was successful by typing from the installation directory root:

Author: Our tests with Moonbeam are using V14.1.1 - But we can try bumping this to 17.x and check

```
node_modules/.bin/substrate-api-sidecar --version
```

## Setting up the Substrate API Sidecar {: #setting-up-the-substrate-api-sidecar }

In the terminal that Sidecar will run, export the environmental variable for the WS endpoint of the network. For example: 

=== "Dancebox"
    ```
    export SAS_SUBSTRATE_URL=wss://fraa-dancebox-rpc.a.dancebox.tanssi.network
    ```

> === "Dancebox EVM ContainerChain"

Author: I don't think this is necessary. I'm not sure how the endpoint structure is going to be. I can add a note saying that you should replace this with your ContainerChain endpoint

The following code samples will demonstrate how to listen to both native token transfers, sent via Substrate or Ethereum API, and ERC-20 token transfers sent via the Ethereum API, using Substrate API Sidecar. Transfers via the Ethereum API are only applicable to EVM ContainerChains.

### Native Token Transfers { #native-token-transfers }

Both Tanssi, non-EVM ContainerChains and EVM ContainerChains can perform Substrate-based native token balance transfers.

The following code snippet uses the Axios HTTP client to query the Sidecar endpoint [`/blocks/head`](https://paritytech.github.io/substrate-api-sidecar/dist/){target=_blank} for the latest finalized block, and then decodes the block for the `from`, `to`, `value`, `tx hash` and `transaction status` of native token transfers at both the EVM and Substrate API level. 

```typescript
--8<-- 'code/substrate-api/sidecar-transfer.ts'
```

### ERC-20 Token Transfers {: #erc-20-token-transfers }

Events emitted by smart contracts such as an ERC-20 token contract deployed on EVM ContainerChains can be decoded from Sidecar block JSON objects. The nesting structure is as following:

> ```JSON

Reviewer: ```suggestion
```
```

Reviewer: I'd delete the JSON type of code. It renders the text in a funny way.
(Applies to every tree type structure description in the doc)

```
GET /pallets/baseFee/storage/baseFeePerGas?at={blockId}
```

Read as a JSON object, the relevant nesting structure is as follows:

```JSON
RESPONSE JSON Storage Object:
    |--at
        |--hash
        |--height
    |--pallet
    |--palletIndex
    |--storageItem
    |--keys
    |--value
```

The relevant data will be stored in the `value` key of the JSON object. This value is a fixed point data type, hence the real value is found by dividing the `value` by the decimals.

### GasPrice, MaxFeePerGas, and MaxPriorityFeePerGas {: #gasprice-maxfeepergas-maxpriorityfeepergas }

The `GasPrice` is used to specify the gas price of legacy transactions prior to [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559){target=_blank}. The `MaxFeePerGas` and `MaxPriorityFeePerGas` were both introduced in EIP-1559 alongside the `BaseFee`. The `MaxFeePerGas` defines the maximum fee permitted to be paid per unit of gas and is the sum of the `BaseFee` and the `MaxPriorityFeePerGas`. The `MaxPriorityFeePerGas` is the maximum priority fee configured by the sender of a transaction that is used to incentive the prioritization of a transaction in a block.

> Although EVm ContainerChains are Ethereum-compatible, they are also Substrate-based chains at its core, and priorities work differently in Substrate than in Ethereum. In Substrate, transactions are not prioritized by gas price. To address this, EVM ContainerChains uses a modified prioritization system that reprioritizes Substrate transactions using an Ethereum-first solution. A Substrate transaction still goes through the validity process, where it is assigned transaction tags, longevity, and a priority. The original priority is then overwritten with a new priority based on the transaction's fee per gas, which is derived from the transaction's tip and weight. If the transaction is an Ethereum transaction, the priority is set according to the priority fee.

Reviewer: ```suggestion
Although EVM ContainerChains are Ethereum-compatible, they are also Substrate-based chains at their core, and priorities work differently in Substrate and in Ethereum. In Substrate, transactions are not prioritized by gas price. To address this, EVM ContainerChains uses a modified prioritization system that reprioritizes Substrate transactions using an Ethereum-first solution. A Substrate transaction still goes through the validity process, where it is assigned transaction tags, longevity, and priority. The original priority is then overwritten with a new priority based on the transaction's fee per gas, which is derived from the transaction's tip and weight. If the transaction is an Ethereum transaction, the priority is set according to the priority fee.
```
